server/src/main/java/org/elasticsearch/watcher/ResourceWatcherService.java

```diff
@@ -75,32 +76,49 @@ public enum Frequency {
 
     private final boolean enabled;
 
+    private final Scheduler.SafeScheduledThreadPoolExecutor executor;
+
     final ResourceMonitor lowMonitor;
     final ResourceMonitor mediumMonitor;
     final ResourceMonitor highMonitor;
 
-    private final Cancellable lowFuture;
-    private final Cancellable mediumFuture;
-    private final Cancellable highFuture;
+    private final Scheduler.Cancellable lowFuture;
+    private final Scheduler.Cancellable mediumFuture;
+    private final Scheduler.Cancellable highFuture;
 
     public ResourceWatcherService(Settings settings, ThreadPool threadPool) {
         this.enabled = ENABLED.get(settings);
 
-        TimeValue interval = RELOAD_INTERVAL_LOW.get(settings);
-        lowMonitor = new ResourceMonitor(interval, Frequency.LOW);
-        interval = RELOAD_INTERVAL_MEDIUM.get(settings);
-        mediumMonitor = new ResourceMonitor(interval, Frequency.MEDIUM);
-        interval = RELOAD_INTERVAL_HIGH.get(settings);
-        highMonitor = new ResourceMonitor(interval, Frequency.HIGH);
         if (enabled) {
-            final var executor = threadPool.generic();
+            TimeValue interval = RELOAD_INTERVAL_LOW.get(settings);
+            lowMonitor = new ResourceMonitor(interval, Frequency.LOW);
+            interval = RELOAD_INTERVAL_MEDIUM.get(settings);
+            mediumMonitor = new ResourceMonitor(interval, Frequency.MEDIUM);
+            interval = RELOAD_INTERVAL_HIGH.get(settings);
+            highMonitor = new ResourceMonitor(interval, Frequency.HIGH);
+            executor = new Scheduler.SafeScheduledThreadPoolExecutor(1, (runnable, exec) -> {
```

Comment:
I don't think we need a `ScheduledThreadPoolExecutor` here, could we just use a regular `EsThreadPoolExecutor` instead? I'd rather we customized the rejection behaviour by making `ResourceMonitor` an `AbstractRunnable` and overriding its `onRejection`, although I think this would need some changes to `ThreadPool#scheduleWithFixedDelay` too.

Reply:
Thanks, I'll take a look at doing that. I wasn't sure what our best practices are these days for things like this, so I'm definitely open to taking a different tack. I saw the fix, but it's still causing a bit of log spam due to the long task durations, so I think doing it on a dedicated thread (or at least not blocking a general-use threadpool about it) would be best.

Reply:
Hmm what log spam are you seeing? I wouldn't expect any warnings about long-running tasks on `generic`.

Reply:
I did a very silly thing; when I checked that this was still a problem... I did not filter the logs by version. If I filter out versions below 8.9.0 (where the fix you reference was released), there is no log spam. I'm going to blame it on getting too used to serverless logs and everything being a very current version all the time. I think I'm just going to close this (and the referenced issue) as I don't think there's enough of a problem remaining to justify more effort in the short term; apologies for the noise.